### ISSUE TITLE
Loss Calculation: MultilabelSoftMarginLoss on top of Softmax Output?

### ISSUE THREAD

**Issue by varshanth** [YapengTian/AVE-ECCV18#15]
Hi,

Thanks for your work. I had an enquiry as to how you are calculating the loss during training. The nn.MultilabelSoftMarginLoss operates on your softmax output. The nn.MultilabelSoftMarginLoss according to PyTorch source code performs BCELoss over the 2nd dimension (https://github.com/pytorch/pytorch/blob/f2af07d7f6ca775baf3abfb058cc5bb4aac7cc3a/torch/nn/functional.py#L2733) . In this, case, if your target is of shape (N, 10, 29), you will be applying a sigmoid operation (from the BCE) on your original softmax output (https://github.com/YapengTian/AVE-ECCV18/blob/master/models.py#L79), and summing the loss across the sequence length (10) instead of on the batch dimension. 

My questions are:

1) Does applying a sigmoid (implicitly through the nn.MultilabelSoftMarginLoss) on a softmax output have some theoretical foundation? The softmax output would never accumulate a negative value and hence, your sigmoid output can never go below 0.5.

2) The use of nn.MultilabelSoftMarginLoss, on (N, 10, 29) output will average the loss over the sequence (10) dimension and not the batch dimension (N).

I am not sure if I am missing something obvious which may involve a deeper understanding of the underlying mechanics of PyTorch. Request you to please explain on the above 2 points.

Thanks a lot for your help!

**Comment by YapengTian** [comment]
Thanks for pointing it out. I previously misused the loss. But with the irregular usage, the model can still be trained. 

Let's look at the loss. From the implementation, we can found that     loss = -(target * logsigmoid(input) + (1 - target) * logsigmoid(-input)). Remember that we used one-hot encoding. When target = 0, sigmoid(-input) will be smaller than 0.5.  The loss will be large if the input is a large value (softmax result). So, the loss will push the negative class to have smaller softmax output (input). If target = 1, minimizing the loss will push the "input" larger. So, the loss with softmax output can still work in the correct direction. For the dimension issue, we can see that it will first do sum over dim=1 then do mean to get the final scalar rather than first sum over dim=2 then do mean. But, it will not affect the optimization much. 

But, anyway, thanks for pointing it out. The irregular way can work but we should use a regular way (softmax + multi-class cross-entropy). Any thoughts? Thanks!

**Comment by varshanth** [comment]
Thanks a lot for your reply. Here are my thoughts:
1) Since the gradient is calculated with respect to the loss, the model output is first going through a sigmoid operation and hence the gradient will be subjected to the derivative of that first. As we know, the gradient will diminish and hence, may take longer to train and may not achieve optimal performance
2) After working out the maths, it seems that we will be penalizing the true positives and true negatives as well, whereas we will be penalizing the false positives and false negatives less (since the range input to the sigmoid is technically gated by the softmax).
3) Suppressing the gradient calculation by 10 instead of the batch size would technically render the batch size hyperparameter useless (unless for GPU fitting purposes) since we are not averaging the batch size but by a fixed value of 10.
4) I am currently testing out the CrossEntropyLoss by flattening the first 2 dims (batch and seq length) and then multiplying the resultant averaged loss (denominator =batch * seq) by seq to correct the behaviour.

My current observations are that with a LR (1e-4), model converges in just 6-10 epochs, which validates above points 1) and 2)

Your code base is a great started code for work on the dataset, so I am reusing your code with my own base network. For my use case, the LR tweaking is a bit harder now since the model is converging so fast.

Please let me know if there is a cleaner way to achieve this or if you have any other suggestions.

Thanks!

**Comment by YapengTian** [comment]
Yes, I agree that shrinking the outputs might reduce training speed and performance (for performance, I am not 100% sure.) It is a little bit similar to a trick that adds a temperature parameter in softmax function for knowledge distillation (https://medium.com/analytics-vidhya/knowledge-distillation-dark-knowledge-of-neural-network-9c1dfb418e6a).  

It is great to know that the model can converge in just several epochs.  I think your implementation is good! One thing you could further try is to add weights for different classes in the loss since the data set is not fully balanced for different event categories. But I can not ensure that it must work but at least worth trying. Thanks for the discussions!

**Comment by varshanth** [comment]
Oddly, the performance of your implementation seems to be more stable. I am having a hard time tweaking the LR for the plain cross-entropy version! It's a nice trick, I would love to hear if you have any additional intuition on that. 

With respect to the weights, definitely, I will try. It's good that nn.MultiLabelSoftMarginLoss accepts weights as parameter. Thanks to you as well, it's good to get prompt clarifications from the authors like this!

**Comment by YapengTian** [comment]
Wow, it is interesting! You might add a temperature parameter in softmax function similar to knowledge distillation (https://medium.com/analytics-vidhya/knowledge-distillation-dark-knowledge-of-neural-network-9c1dfb418e6a). I think the used strategy is somehow similar to this trick. I am also very pleased to have discussions with you.
